### PR TITLE
Navigator: add support for exit animation

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Part of #59418
Requires #64786 to be merged first

Add exit animations to `Navigator` screen

## Why?

An exit animation can help the user understand the transition to a new screen better — and gives an additional sense of polish to the UI.

The animation specs (durations, delays, easing functions) are also updated following the latest motion guidelines (cc @nick-a8c)

## How?

  - Added a `useScreenAnimatePresence` hook which takes care of managing the screen's animation lifecycle
  - Use CSS transitions to animate the screen
  - Use `data-` attributes to set animation state on the DOM element, which is used to trigger the CSS transitions
  - Use CSS grid on the provider to force all screens to overlap, while also pushing the height of the container
  - Always render the screen root element, while conditionally rendering its contents
  - Updated tests to completely skip animations, that were causing errors and general flakiness
  - Added a note in the README about setting the wrapper height (related conversation in https://github.com/WordPress/gutenberg/pull/64798#discussion_r1733235033)

## Testing Instructions

- Make sure that the animations are always applied to the pair of screens (the one transitioning in and the one transitioning out), both when navigating to a new screen and back to the parent screen;
- Make sure that the animations work as expected on all supported browsers;
- Make sure that the animations are not performed when enabling "reduced motion" browser/OS settings

## Screenshots or screencast 

https://github.com/user-attachments/assets/bee8f934-0f0c-4df1-93ae-5510b2f4abd9

10x slow-motion:

https://github.com/user-attachments/assets/d802adc5-6b4f-4803-b8c0-26a421e618cf

### CODE REVIEW

**Comment by jasmussen** [comment]
Nice. Is it possible to test this in the site editor? It's not clear that the new animation from this branch is applied there:

nav

Here's trunk for comparison:

trunk

**Comment by ciampo** [comment]
It appears that the site editor sidebar doesn't use `Navigator` anymore — looking at the file history, the change happened in https://github.com/WordPress/gutenberg/pull/60466. I was on parental leave at the time, but it seems like the main reason was to use the global site-editor router instead of having to coordinate with `Navigator`'s internal router.

`Navigator` is still being used in the Global Styles sidebar, though.

File: packages/components/src/navigator/navigator-screen/component.tsx
Comment (ciampo) [inline packages/components/src/navigator/navigator-screen/component.tsx]: I may consider adding a timeout to make sure that, if for whatever reason the correct `animationend` event doensn't fire, the timeout is going to still cause the screen to be correctly from the DOM.

Comment (ciampo) [reply packages/components/src/navigator/navigator-screen/component.tsx]: Added in https://github.com/WordPress/gutenberg/pull/64777/commits/0ae892a27f7ea9dee23efd8fd968847dec0fe078

**Review by mirka — COMMENTED** [review]
Animating presence will likely be a recurring need. How can we abstract this into something encapsulated and reusable, so we don't have to add in a bunch of ad hoc animation coordinating logic into a component every time we need an entry/exit animation?

Not immediately sure if we should use a library like `react-transition-group` ([maybe](https://bundlephobia.com/package/react-transition-group@4.4.5) 1.7kB gzipped if we only needed the `Transition` component), or if we want to maintain our own for an even smaller size. Thoughts?

**Comment by ciampo** [comment]
> Animating presence will likely be a recurring need

That's true, although in many cases, animating presence [can be done with some recent CSS features](https://developer.chrome.com/blog/entry-exit-animations). But those can't be used with `Navigator` because we remove a `Screen`'s markup when the screen gets deselected.

> Not immediately sure if we should use a library like react-transition-group ([maybe](https://bundlephobia.com/package/react-transition-group@4.4.5) 1.7kB gzipped if we only needed the Transition component), or if we want to maintain our own for an even smaller size.

I'm conflicted — on one end, I don't like to reinvent the wheel if the problem has been solved well by third-parties. On the other, we're trying to [move away from `framer-motion`](https://github.com/WordPress/gutenberg/issues/60975) for smaller bundle size and better performance (using CSS native features as much as possible).

If we used `react-transition-group`, we'd need to use the `CSSTransition` component (`2.6kB`), since the `Transition` component is just the abstract logic.

Also, this specific type of animation will be much easier when [view transitions](https://developer.mozilla.org/en-US/docs/Web/API/View_Transitions_API) become more advanced and better supported in the next few years.

> How can we abstract this into something encapsulated and reusable, so we don't have to add in a bunch of ad hoc animation coordinating logic into a component every time we need an entry/exit animation?

I wonder if doing so would represent a form of early optimization. Maybe we can keep `Navigator` as-is, and re-discuss this topic if/when we have more real use cases for animating presence, so that we can better understand what type solution we'd need to implement?

File: packages/components/src/navigator/navigator-screen/component.tsx
Comment (mirka) [inline packages/components/src/navigator/navigator-screen/component.tsx]: I found a [`useReducedMotion` hook](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-compose/#usereducedmotion) in wp-compose.

**Review by mirka — COMMENTED** [review]
> I wonder if doing so would represent a form of early optimization. Maybe we can keep `Navigator` as-is, and re-discuss this topic if/when we have more real use cases for animating presence, so that we can better understand what type solution we'd need to implement?

I'm not familiar enough to make a confident judgement in either direction, but I trust you on your assessment after working on it.

I guess what immediately makes me a bit uncomfortable though is how the animation logic is interspersed in the component code. Do you think it would be worth the effort to extract most or all of the animation logic into a single hook? No regard for reusability at the moment, but just to isolate all the animation concerns into a single place. I'm already a bit afraid of touching this code, and I think separating the concerns would help a lot to ease that sense of complexity.

**Comment by ciampo** [comment]
Rebased to include all recent `Navigator` changes. I will look at addressing the remaining feedback by the end of the week

File: packages/components/src/navigator/navigator-provider/README.md
Comment (tyxla) [inline packages/components/src/navigator/navigator-provider/README.md]: Ah, this is a bummer. Looks like something the component should be able to do on its own. Is there a reason why we can't automate this, using a resize observer for example?

File: packages/components/src/navigator/navigator-screen/component.tsx
Comment (ciampo) [reply packages/components/src/navigator/navigator-screen/component.tsx]: Good suggestion! Done in https://github.com/WordPress/gutenberg/pull/64777/commits/15fe373ca8e090494f5190874ec8daad91f12e6f

**Comment by ciampo** [comment]
> I guess what immediately makes me a bit uncomfortable though is how the animation logic is interspersed in the component code. Do you think it would be worth the effort to extract most or all of the animation logic into a single hook? No regard for reusability at the moment, but just to isolate all the animation concerns into a single place.

This is a great suggestion, I definitely agree. I pushed a couple of commits and things are looking much more clear :)

File: packages/components/src/navigator/navigator-screen/component.tsx
Comment (ciampo) [inline packages/components/src/navigator/navigator-screen/component.tsx:86]: There are no runtime changes re. focus restoration, only variable renaming and some prop destructuring that happened outside the `useEffect` hook

File: packages/components/src/navigator/test/index.tsx
Comment (tyxla) [inline packages/components/src/navigator/test/index.tsx]: Isn't it a better bet to test with all animations enabled? After all, there might be bugs that are only encountered with animations on.

**Review by tyxla — COMMENTED** [review]
I still need to test and review this thoroughly, but I'm submitting some early, naive feedback.

File: packages/components/src/navigator/navigator-screen/component.tsx
Comment (tyxla) [inline packages/components/src/navigator/navigator-screen/component.tsx]: Some of the comments seem a bit redundant since the code is self-explicatory. Were you planning on cleaning them up once you're done?

File: packages/components/src/navigator/test/index.tsx
Comment (ciampo) [reply packages/components/src/navigator/test/index.tsx]: I don't think `js-dom` supports animations well — `testing-library` has even an [example](https://testing-library.com/docs/example-react-transition-group/#mock) of how to mock `React Transition Group`.

File: packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts
Comment (tyxla) [inline packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts:51]: Lots of good work here, but I'm wondering if we're trying too hard not to use `framer-motion`. When it's about a simple animation, using CSS makes lots of sense. But when we need to reinvent the wheel in cases like animating presence, I feel like the best bet would be to use `framer-motion`. Bundle size is important, but DX and UX are more important, so we might have to make compromises. Especially given that we're looking to build more complex motion and animations for many of the components and flows.

File: packages/components/src/navigator/navigator-provider/README.md
Comment (ciampo) [reply packages/components/src/navigator/navigator-provider/README.md]: I pushed some more work that should greatly improve this aspect.

File: packages/components/src/navigator/navigator-screen/component.tsx
Comment (ciampo) [reply packages/components/src/navigator/navigator-screen/component.tsx]: Removed obvious comments

**Comment by ciampo** [comment]
I pushed some further updates to the logic, making it (hopefully) clearer and making more resilient to UI "jumping" while switching across screens.

I guess the conversation started in https://github.com/WordPress/gutenberg/pull/64777#discussion_r1736202491 will inform the future of this PR — whether we go ahead with the current approach, or if we revert back to using a higher-level animation library.

File: packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts
Comment (mirka) [reply packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts:51]: I definitely feel _better_ about it after the refactor into a designated hook, but also it’s still unknown how bug-free the current code is, and how many other components will need to animate presence.

My instinct is that it could still make sense to migrate from Framer Motion to a lighter library like React Transition Group. But then again, if it's a substantial effort to migrate from FM to RTG, maybe even that isn't worth it. 

I’d love to hear from @ciampo and @daniguardiola about how they currently assess the overall cost/benefit around these rewrites. Let's make sure the benefits are still big enough to justify our rewrite efforts and potential increase in maintenance cost.

Comment (mirka) [reply packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts:51]: As for this particular PR, I'm personally comfortable enough with the logic isolation now to merge it as an incremental improvement over trunk. Is it scalable/sustainable? Maybe not. But it seems like we could swap it out without too much difficulty if we ultimately decide to use a library.

Comment (tyxla) [reply packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts:51]: I don't have particularly strong feelings about blocking this particular PR in favor of `framer-motion`, but I share the sentiment on scalability, the extra maintenance cost, and moving in that direction in the future based on the amount of custom work and wheel rediscovery we need to do. It's worth considering doing more specific measurements to ensure that a refactor away from `framer-motion` actually yields noticeable results. We should also consider the opportunity cost - what we could have built/shipped if we reused `framer-motion` instead of rediscovering it and maintaining the extra layer of complexity.

**Comment by ciampo** [comment]
Pausing the effort on this PR for now for a couple of reasons:

- I'd like to have agreement on [whether we should use a 3rd party library or not](https://github.com/WordPress/gutenberg/pull/64777#discussion_r1736202491)
- some animations in the global styles sidebar are _slightly_ glitchy for reasons I yet don't understand (especially when going from the Typography screen to the individual Element (ie. Text, Headings, Lists...) typography screen. I won't spend further time debugging the issue if we're not going to go down this PR's route

**Comment by ciampo** [comment]
I had a couple of ideas going around in my mind, I tested them and they seem to improve the logic.

I also discovered this small "transition state machine" library that could delegate some logic to a third party, if we even think it's worth exploring. https://www.npmjs.com/package/react-transition-state

File: packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts
Comment (jsnajdr) [inline packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts]: `animationTimeout` can be a local variable in the effect function, no need for an external ref.

Also, should this fallback timeout also change the state from `ANIMATING_IN` to `IN`? The `onScreenAnimationEnd` does that. If I understand correctly, the idea is that if the `animationEnd` callback doesn't run for some reason, the fallback callback will do its job anyway after some timeout.

Comment (ciampo) [reply packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts]: Yes, that's correct — I forgot to update the timeout in the last refactor

Comment (ciampo) [reply packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts]: https://github.com/WordPress/gutenberg/pull/64777/commits/ce9e38acbc9ddb274bcfb6ecf61277fa1e8987bf

**Comment by jsnajdr** [comment]
Here's how I understand the concept:

The `NavigatorProvider` no longer renders just one matching `NavigatorScreen`, but it can render two or even more. One is the real active one, the others are there only for effect. We could build a navigator that displays all parent screens using a `wasMatch` array, like this Apple Time Machine history UI 🙂

ea02d34d-0309-4109-819a-e05eb5087624

So, that determines the DOM structure and the baseline styles for navigator, even without any animations. The matching screens are displayed side-by-side, or they are all `display:none` except one, or are displayed as something else. The styles must be able to show the active screen as the active one and the other screens as inactive.

I don't like the `setWrapperHeight` helper very much. It feels very hacky. Instead, the base navigator styles should be able to display multiple screens at once without any helpers. When I remove `setWrapperHeight` from the code, I don't see any ugly effects in the Storybook examples.

There can be a behavior that the `wasMatch` value is reset to `false` after some timeout and the inactive screen is unmounted. That will be useful when the exit animation is implemented. But it's an optimization, not an essential feature. The navigator still works 100% when the `wasMatch` screen is rendered indefinitely.

When the base DOM structure and the base styles are established, the animations are merely an icing on the cake. The transition from active screen to inactive and vice versa are animated. Can be easily done with CSS animations.

I think that when you think about the PR as i) solid base DOM structure and styles + ii) simple animations on top, it's easier to understand.

And if we did this with `framer-motion`, it doesn't give us much extra, does it? Its `AnimatePresence` helper is useful, but that's all.

**Comment by ciampo** [comment]
> When I remove setWrapperHeight from the code, I don't see any ugly effects in the Storybook examples.

That's because, in the storybook examples, the `NavigatorProvider` component has a `height: calc(100vh - 2rem)` style applied.

If you remove the set `height` and remove `setWrapperHeight` too, you should notice that the screen animating out will have its contents cutoff if the screen animating in is not as tall.

Here is a video comparison (I slowed down the animation and added a background gradient to make things more evident)

| Without `setWrapperHeight` | With `setWrapperHeight` |
|---|---|
| <video src="https://github.com/user-attachments/assets/6d4d30af-1d0b-4a88-a8c9-300823ca2752" /> | <video src="https://github.com/user-attachments/assets/6658001e-2698-40cd-a2d8-02e7f0481825" /> |

Having said that, this comment made me reflect:

> I don't like the setWrapperHeight helper very much. It feels very hacky. Instead, the base navigator styles should be able to display multiple screens at once without any helpers.

And I realised that, instead of setting `position: absolute` on the screen transitioning out (which then causes the `NavigatorProvider` to jump, and the screen animating out to be possible cut off), we can use `display: grid` and have both screens (the one transitioning in, and the one transitioning out) overlap in the same grid cell, meaning that the tallest one will push the cell's height without the need for `setWrapperHeight`.

I applied this approach in [`8bca499` (#64777)](https://github.com/WordPress/gutenberg/pull/64777/commits/8bca499c4ed77bf7b45077ad380a8357a5424906)

**Comment by ciampo** [comment]
> Here's how I understand the concept:
> 
> The `NavigatorProvider` no longer renders just one matching `NavigatorScreen`, but it can render two or even more. One is the real active one, the others are there only for effect. We could build a navigator that displays all parent screens using a `wasMatch` array, like this Apple Time Machine history UI 🙂
> 
> So, that determines the DOM structure and the baseline styles for navigator, even without any animations. The matching screens are displayed side-by-side, or they are all `display:none` except one, or are displayed as something else. The styles must be able to show the active screen as the active one and the other screens as inactive.


If I understand correctly, this is not too different from how `Navigator` works.

The `NavigatorProvider` keeps all the logic, and lets every screen know what's the selected screen ID. So, every `NavigatorScreen` is always rendered (in the react sense).

Each screen, then, decides independently if its contents should be rendered in the DOM, and handles focus if/when needed.

> When the base DOM structure and the base styles are established, the animations are merely an icing on the cake. The transition from active screen to inactive and vice versa are animated. Can be easily done with CSS animations.
> I think that when you think about the PR as i) solid base DOM structure and styles + ii) simple animations on top, it's easier to understand.

Yup, that's basically what's happening with the `useScreenAnimatePresence` logic. And another good demonstration of this separation is that, with the latest changes, this PR didn't have to make any changes to the `NavigatorProvider` (where all the orchestration logic is).

The hook is basically a state machine that goes through the "animation lifecycle" of a navigator screen:

```
initial     →           animating in          →        visible
                             ↑                            ↓
                   removed from the DOM       ←      animating out
```     

> And if we did this with `framer-motion`, it doesn't give us much extra, does it? Its `AnimatePresence` helper is useful, but that's all.

Framer motion would basically replace the `useScreenAnimatePresence` hook, but everything else would still need to be kept around.

**Comment by ciampo** [comment]
On a separate note, the bug that I mentioned before in the global styles sidebar continues to persist. I tried to slow down the animations in the dev tools, but that causes the bug not to be reproducible. I had to use my phone to record a slow-motion video of the transition:


https://github.com/user-attachments/assets/9b9972b0-c570-48e3-81cf-1703ed771eb8

It's as if the paragraph screen was selected, then deselected for a brief moment, and then selected again.

I will look into it, as I'm not sure yet if this is something related to how `Navigator` is used in global styles, or if this is an internal bug to `Navigator`. And I'm not even sure if this is something that happens on `trunk` already (and is just not visible because there's no exit animation).

So far, I could only reproduce this glitch when navigating from the list of blocks to the specific block — and nowhere else in global styles.

**Comment by ciampo** [comment]
So!

I spent a lot more time than I would have liked to debug [the glitch in the global styles sidebar](https://github.com/WordPress/gutenberg/pull/64777#issuecomment-2334381574), and couldn't really get to the bottom of it. Somehow, the CSS animation gets re-triggered,and I wasn't able to understand the reason (not sure if @jsnajdr wants to attempt to debug it too).

I therefore decided to apply a different animation strategy, **using CSS transitions instead of animations** since they don't restart if the underlying styles get re-applied. I tested the new approach in Storybook and in the editor, both Global Styles sidebar and preferences modal (on mobile viewport) and it seems to work well.

@WordPress/gutenberg-components , do we think we can polish + merge the current approach? (I also made a few changes to the global styles sidebar code, which I will extract to separate PRs if we go forward with the current set of changes)

File: packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts
Comment (ciampo) [inline packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts]: This comment is out of date, I'll remove it in a future commit

**Comment by ciampo** [comment]
With https://github.com/WordPress/gutenberg/pull/65494 merged, the glitch reported above should be fixed (thank you @jsnajdr 🙏 )

I rebased, restored the CSS animations based approach, and added a few tweaks (like moving to `data-*` attributes, instead of dynamic emotion styles).

I'm going to extract the changes to the `@wordpress/edit-site` package to separate PRs, but in the meantime this PR is ready to be tested again.

**Comment by ciampo** [comment]
With https://github.com/WordPress/gutenberg/pull/65494 merged, the [glitch](https://github.com/WordPress/gutenberg/pull/64777#issuecomment-2334381574) reported above should be fixed (thank you @jsnajdr 🙏 )

I rebased, restored the CSS animation- based approach, and added a few tweaks (like moving to `data-*` attributes, instead of dynamic emotion styles).

I'm going to extract changes not related to the exit animation to separate PRs (#65522, #65523, #65527), but in the meantime, this PR is ready to be tested again.

**Comment by ciampo** [comment]
> With #65494 merged, the [glitch](https://github.com/WordPress/gutenberg/pull/64777#issuecomment-2334381574) reported above should be fixed

Since that PR [was reverted](https://github.com/WordPress/gutenberg/pull/65533) I'd imagine the glitch would be back when rebasing on top of latest trunk 😭

**Comment by jsnajdr** [comment]
> Since that PR was reverted I'd imagine the glitch would be back when rebasing on top of latest trunk

I'm trying again in #65564, this time with a much better and simpler implementation 🙂

**Comment by tyxla** [comment]
Sounds like we should mark this one as blocked until #65564 lands.

**Comment by ciampo** [comment]
With https://github.com/WordPress/gutenberg/pull/65564, https://github.com/WordPress/gutenberg/pull/65522, https://github.com/WordPress/gutenberg/pull/65523, and https://github.com/WordPress/gutenberg/pull/65527 merged, I rebased this PR, solved conflicts and remove all changes external to `Navigator`.

It tested well on my machine all all browsers

**Review by jsnajdr — APPROVED** [review]
I don't have any further feedback, everything has been addressed. The glitch with Typography panel is solved. I think this is ready to 🚢

**Comment by t-hamano** [comment]
This PR seems to have caused problems with the shadow preset panel and font size preset panel.

Looking at the browser error log, it might be a problem with the implementation of the shadow/font size preset, but could you take a look at the following issue?

- https://github.com/WordPress/gutenberg/issues/65711

**Comment by ciampo** [comment]
> This PR seems to have caused problems with the shadow preset panel and font size preset panel.
> 
> Looking at the browser error log, it might be a problem with the implementation of the shadow/font size preset, but could you take a look at the following issue?
> 
> * [Shadow/Font size preset panel crashes the editor #65711](https://github.com/WordPress/gutenberg/issues/65711)

Thank you for flagging — left a reply in the issue itself.

**Comment by jeffpaul** [comment]
@nick-a8c would you mind sharing your WordPress.org username so we can ensure you're properly credited in the 6.7 release next week?

**Pull request merged**
